File: cassini.tm/src/main/java/com/cassinisys/tm/service/AccommodationService.java
package com.cassinisys.tm.service;

import com.cassinisys.platform.exceptions.ResourceNotFoundException;
import com.cassinisys.platform.service.core.CrudService;
import com.cassinisys.platform.service.core.PageableService;
import com.cassinisys.tm.filtering.AccommodationCriteria;
import com.cassinisys.tm.filtering.AccommodationPredicateBuilder;
import com.cassinisys.tm.model.*;
import com.cassinisys.tm.model.dto.KeyAndNumber;
import com.cassinisys.tm.repo.AccommodationRepository;
import com.cassinisys.tm.repo.AccommodationSuitRepository;
import com.cassinisys.tm.repo.BedRepository;
import com.cassinisys.tm.repo.SuitRepository;
import com.mysema.query.types.Predicate;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.PageRequest;
import org.springframework.data.domain.Pageable;
import org.springframework.data.domain.Sort;
import org.springframework.stereotype.Service;
import org.springframework.transaction.annotation.Transactional;

import java.util.ArrayList;
import java.util.List;

import static com.google.common.base.Preconditions.checkNotNull;

/**
 * Created by CassiniSystems on 03-08-2016.
 */
@Service
@Transactional
public class AccommodationService implements CrudService<TMAccommodation, Integer>, PageableService<TMAccommodation, Integer> {

    @Autowired
    private AccommodationRepository accommodationRepository;

    @Autowired
    private AccommodationSuitRepository accommodationSuitRepository;

    @Autowired
    private BedRepository bedRepository;

    @Autowired
    private AccommodationPredicateBuilder accommodationPredicateBuilder;


    @Autowired
    private SuitRepository suitRepository;

    @Override
    public TMAccommodation create(TMAccommodation accommodation) {
        checkNotNull(accommodation);
        return accommodationRepository.save(accommodation);
    }

    @Override
    public TMAccommodation update(TMAccommodation accommodation) {
        checkNotNull(accommodation);
        checkNotNull(accommodation.getId());
        return accommodationRepository.save(accommodation);
    }

    @Override
    public void delete(Integer id) {
        checkNotNull(id);
        TMAccommodation accommodation = accommodationRepository.findOne(id);
        if (accommodation == null) {
            throw new ResourceNotFoundException();
        }
        accommodationRepository.delete(id);
    }

    @Override
    public TMAccommodation get(Integer id) {
        checkNotNull(id);
        TMAccommodation accommodation = accommodationRepository.findOne(id);
        if (accommodation == null) {
            throw new ResourceNotFoundException();
        }
        return accommodation;
    }

    @Override
    public List<TMAccommodation> getAll() {
        return accommodationRepository.findAll();
    }

    @Override
    public Page<TMAccommodation> findAll(Pageable pageable) {
        checkNotNull(pageable);
        return accommodationRepository.findAll(pageable);
    }

    public Page<TMAccommodation> findAll(AccommodationCriteria criteria, Pageable pageable) {
        if (pageable.getSort() == null) {
            pageable = new PageRequest(pageable.getPageNumber(), pageable.getPageSize(),
                    new Sort(new Sort.Order(Sort.Direction.DESC, "name")));
        }
        Predicate predicate = accommodationPredicateBuilder.build(criteria, QTMAccommodation.tMAccommodation);
        Page<TMAccommodation> accs = accommodationRepository.findAll(predicate, pageable);
        return accs;
    }

    @Transactional(readOnly = true)
    public List<TMAccommodation> findMultipleAccommodations(List<Integer> ids) {
        return accommodationRepository.findByIdIn(ids);
    }

    public TMAccommodationSuit createAccommodationSuit(TMAccommodationSuit accommodationSuit) {
        return accommodationSuitRepository.save(accommodationSuit);
    }

    public TMBed createSuitBed(TMBed tmBed) {
        return bedRepository.save(tmBed);
    }

    public TMBed updateSuitBed(TMBed tmBed) {
        return bedRepository.save(tmBed);
    }

    public TMAccommodationSuit updateAccommodationSuit(TMAccommodationSuit accommodationSuit) {
        checkNotNull(accommodationSuit);
        checkNotNull(accommodationSuit.getSuitId());
        return accommodationSuitRepository.save(accommodationSuit);
    }

    public void deleteAccommodationSuit(Integer suitId) {
        checkNotNull(suitId);
        TMAccommodationSuit accommodationSuit = accommodationSuitRepository.findOne(suitId);
        if (accommodationSuit == null) {
            throw new ResourceNotFoundException();
        }
        accommodationSuitRepository.delete(accommodationSuit.getSuitId());
    }

     public void deleteBed(Integer bedId) {
        checkNotNull(bedId);
        bedRepository.delete(bedId);
    }

    public TMAccommodationSuit getAccommodationSuit(Integer Id) {
        checkNotNull(Id);
        TMAccommodationSuit accommodationSuit = accommodationSuitRepository.findOne(Id);
        if (accommodationSuit == null) {
            throw new ResourceNotFoundException();
        }
        return accommodationSuit;
    }

    public List<TMAccommodationSuit> getAllSuits() {
        return accommodationSuitRepository.findAll();
    }

    public Page<TMAccommodationSuit> findAllSuits(Pageable pageable) {
        return accommodationSuitRepository.findAll(pageable);
    }

    public List<TMAccommodationSuit> getSuitsByAccommodation(Integer accomId) {

        return accommodationSuitRepository.findByAccommodation(accomId);
    }

    public List<TMBed> getBedsBySuitId(Integer suitId) {
        return bedRepository.findBySuite(suitId);
    }



    public TMSuit createSuitPerson(TMSuit suitPerson) {
        checkNotNull(suitPerson);
        return suitRepository.save(suitPerson);
    }

    public TMSuit updateSuitPerson(TMSuit suitPerson) {
        checkNotNull(suitPerson);
        checkNotNull(suitPerson.getRowId());
        return suitRepository.save(suitPerson);
    }

    public void deleteSuitPerson(Integer id) {
        checkNotNull(id);
        accommodationSuitRepository.delete(id);
    }

    public TMSuit getSuitPerson(Integer Id) {
        checkNotNull(Id);
        TMSuit suitPerson = suitRepository.findOne(Id);
        if (suitPerson == null) {
            throw new ResourceNotFoundException();
        }
        return suitPerson;
    }

    public List<TMSuit> getAllSuitPersons() {
        return suitRepository.findAll();
    }

    public Page<TMSuit> findAllSuitPersons(Pageable pageable) {
        return suitRepository.findAll(pageable);
    }

    public List<TMSuit> getPersonsBySuit(Integer suitId) {

        return suitRepository.findBySuite(suitId);
    }

    public List<KeyAndNumber> getCounts() {
        List<KeyAndNumber> counts = new ArrayList<>();
        counts.add(new KeyAndNumber("accommodations", accommodationRepository.count()));
        counts.add(new KeyAndNumber("suites", accommodationSuitRepository.count()));
        counts.add(new KeyAndNumber("beds", bedRepository.count()));
        return counts;
    }

    public List<KeyAndNumber> getSuiteCounts(Integer suiteId) {
        List<KeyAndNumber> counts = new ArrayList<>();
        counts.add(new KeyAndNumber("occupied", bedRepository.findBySuiteAndAssignedToIsNotNull(suiteId).size()));
        counts.add(new KeyAndNumber("available", bedRepository.findBySuiteAndAssignedToIsNull(suiteId).size()));
        return counts;
    }

    public List<TMBed> getSuiteBedsOccupied(Integer suiteId) {
        return bedRepository.findBySuiteAndAssignedToIsNotNull(suiteId);
    }

    public List<TMBed> getSuiteBedsAvailable(Integer suiteId) {
        return bedRepository.findBySuiteAndAssignedToIsNull(suiteId);
    }

    public List<KeyAndNumber> getBedCounts() {
        List<KeyAndNumber> counts = new ArrayList<>();
        counts.add(new KeyAndNumber("occupied", bedRepository.findByAssignedToIsNotNull().size()));
        counts.add(new KeyAndNumber("available", bedRepository.findByAssignedToIsNull().size()));
        return counts;
    }

    public List<TMBed> getOccupiedBeds() {
        return bedRepository.findByAssignedToIsNotNull();
    }

    public List<TMBed> getAvailableBeds() {
        return bedRepository.findByAssignedToIsNull();
    }

    public TMBed getBedByAssignedTo(Integer personId) {
        return bedRepository.findByAssignedTo(personId);
    }
}

File: cassini.tm/src/main/webapp/app/shared/services/accommodationService.js
define(
    [
        'app/shared/services/services.module',
        'app/assets/bower_components/cassini-platform/app/shared/factories/httpFactory'
    ],
    function (module) {
        module.factory('AccommodationService', AccommodationService);

        function AccommodationService(httpFactory) {
            return {
                createAccommodation: createAccommodation,
                getAccommodation: getAccommodation,
                getAccommodationById: getAccommodationById,
                updateAccommodation: updateAccommodation,
                deleteAccommodation: deleteAccommodation,
                getAllAccommodations: getAllAccommodations,
                getMultipleAccommodations: getMultipleAccommodations,

                createSuit: createSuit,
                createBed: createBed,
                updateBed: updateBed,
                getSuit: getSuit,
                getBedsBySuit:getBedsBySuit,
                updateSuit: updateSuit,
                deleteSuit: deleteSuit,
                deleteBed:deleteBed,
                getAllSuits: getAllSuits,
                getSuitById: getSuitById,
                getSuitsByAccommodation: getSuitsByAccommodation,
                freeTextSearch:freeTextSearch,

                getCounts: getCounts,
                getSuiteCounts: getSuiteCounts,
                getSuiteBedsOccupied: getSuiteBedsOccupied,
                getSuiteBedsAvailable: getSuiteBedsAvailable,
                getBedCounts: getBedCounts,
                getOccupiedBeds: getOccupiedBeds,
                getAvailableBeds: getAvailableBeds,
                getBedByAssignedTo: getBedByAssignedTo
            };

            function freeTextSearch(criteria, pageable) {
                var url = "api/accommodations/freesearch?page={0}&size={1}".format(pageable.page, pageable.size);
                url += "&searchQuery={0}".
                    format(criteria.searchQuery);
                return httpFactory.get(url);
            }

            function getMultipleAccommodations(ids) {
                var url = 'api/accommodations/multiple/' + ids;
                return httpFactory.get(url);
            }

            function getAllAccommodations(pageable) {
                var url = 'api/accommodations?page={0}&size={1}'.format(pageable.page, pageable.size);
                return httpFactory.get(url);
            }

            function getAccommodationById(accId) {
                var url = "api/accommodations/" + accId;
                return httpFactory.get(url);
            }

            function createAccommodation(accomediation) {
                var url = 'api/accommodations';
                return httpFactory.post(url, accomediation);
            }

            function updateAccommodation(accommodations) {
                var url = "api/accommodations/" + accommodations.id;
                return httpFactory.put(url,accommodations );
            }

            function deleteAccommodation(accommodationId) {
                var url = "api/accommodations/" + accommodationId;
                return httpFactory.delete(url);
            }

            function getAccommodation(accommodationId) {
                var url = "api/accommodations/" + accommodationId;
                return httpFactory.get(url);
            }

            function createSuit(accommodationId,suit) {
                var url = "api/accommodations/" + accommodationId + "/suits";
                return httpFactory.post(url, suit);
            }

            function getBedsBySuit(accommodationId ,suitId) {
                var url = "api/accommodations/" + accommodationId + "/suits/" + suitId +"/beds";
                return httpFactory.get(url);
            }

            function createBed(accommodationId ,suitId ,bed) {
                var url = "api/accommodations/" + accommodationId + "/suits/" + suitId +"/beds";
                return httpFactory.post(url,bed);
            }

            function updateBed(accommodationId ,suitId ,bed) {
                var url = "api/accommodations/" + accommodationId + "/suits/" + suitId +"/beds/" + bed.bedId;
                return httpFactory.put(url,bed);
            }

            function getSuitsByAccommodation(accommodationId) {
                var url = "api/accommodations/" + accommodationId + "/suits";
                return httpFactory.get(url);
            }


            function getSuit(suit, accommodationId, suitId) {
                var url = "api/accommodations/" + accommodationId + '/suits' + suitId;
                return httpFactory.get(url, suit);
            }

            function getSuitById(accommodationId, suitId) {
                var url = "api/accommodations/" + accommodationId + "/suits/" +suitId;
                return httpFactory.get(url);
            }

            function updateSuit(suit, accommodationId, suitId) {
                var url = "api/accommodations/" + accommodationId / suit + suitId;
                return httpFactory.put(url, suit);
            }

            function deleteSuit(accommodationId, suitId) {
                var url = "api/accommodations/" + accommodationId +"/suits/" + suitId;
                return httpFactory.delete(url);
            }

            function deleteBed(suitId,bedId) {
                var url = "api/accommodations/"+ suitId +"/beds/"+ bedId;
                return httpFactory.delete(url);
            }

            function getAllSuits() {
                var url = "api/accommodations/suits/all";
                return httpFactory.get(url);
            }

            function getCounts() {
                var url = "api/accommodations/counts";
                return httpFactory.get(url);
            }

            function getSuiteCounts(suiteId) {
                var url = "api/accommodations/suits/" + suiteId + "/counts";
                return httpFactory.get(url);
            }

            function getSuiteBedsOccupied(suiteId) {
                var url = "api/accommodations/suits/" + suiteId + "/occupied";
                return httpFactory.get(url);
            }

            function getSuiteBedsAvailable(suiteId) {
                var url = "api/accommodations/suits/" + suiteId + "/available";
                return httpFactory.get(url);
            }

            function getBedCounts() {
                var url = "api/accommodations/suits/beds/counts";
                return httpFactory.get(url);
            }

            function getOccupiedBeds() {
                var url = "api/accommodations/suits/beds/occupied";
                return httpFactory.get(url);
            }

            function getAvailableBeds() {
                var url = "api/accommodations/suits/beds/available";
                return httpFactory.get(url);
            }

            function getBedByAssignedTo(personId) {
                var url = "api/accommodations/suits/beds/" + personId;
                return httpFactory.get(url);
            }
        }
    }
)
;